Source/WebCore/dom/Element.cpp

```diff
@@ -2735,232 +2735,313 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
+// Specs: https://html.spec.whatwg.org/multipage/dom.html#auto-directionality
+static std::optional<TextDirection> computeTextDirectionOfSlotElement(const HTMLSlotElement& slotElement)
 {
-    Style::PseudoClassChangeInvalidation styleInvalidation(*this, CSSSelector::PseudoClass::Dir, Style::PseudoClassChangeInvalidation::AnyValue);
-    auto effectiveDirection = direction.value_or(TextDirection::LTR);
-    setUsesEffectiveTextDirection(!!direction);
-    if (direction)
-        setEffectiveTextDirection(effectiveDirection);
-    auto updateEffectiveTextDirectionOfShadowRoot = [&](Element& element) {
-        if (RefPtr shadowRootOfElement = element.shadowRoot()) {
-            for (Ref element : childrenOfType<Element>(*shadowRootOfElement))
-                element->updateEffectiveDirectionality(direction);
+    // Specs: If element is a slot element whose root is a shadow root
+    // and element's assigned nodes are not empty, then return the
+    // directionality of this slot.
+    if (!slotElement.isInShadowTree())
+        return computeElementContainedTextAutoDirection(slotElement);
+
+    auto* nodes = slotElement.assignedNodes();
+    if (!nodes)
+        return computeElementContainedTextAutoDirection(slotElement);
+
+    for (auto& child : *nodes) {
+        // Specs: If child is a Text node and its text direction is
+        // not null, then return it.
+        if (child->isTextNode()) {
+            if (auto direction = computeTextDirectionFromText(child->textContent(true)))
+                return direction;
         }
-    };
-    updateEffectiveTextDirectionOfShadowRoot(*this);
-    for (auto it = descendantsOfType<Element>(*this).begin(); it;) {
-        Ref element = *it;
-        if (isValidDirValue(element->attributeWithoutSynchronization(dirAttr))) {
-            it.traverseNextSkippingChildren();
-            continue;
+
+        if (auto* element = dynamicDowncast<Element>(child.get())) {
+            // Specs: If child direction is not null, then return child direction.
+            if (auto direction = element->computeAutoTextDirection())
+                return direction;
         }
-        updateEffectiveTextDirectionOfShadowRoot(element);
-        Style::PseudoClassChangeInvalidation styleInvalidation(element, CSSSelector::PseudoClass::Dir, Style::PseudoClassChangeInvalidation::AnyValue);
-        element->setUsesEffectiveTextDirection(!!direction);
-        if (direction)
-            element->setEffectiveTextDirection(effectiveDirection);
-        it.traverseNext();
     }
+
+    return std::nullopt;
 }
 
-void Element::adjustDirectionalityIfNeededAfterChildrenChanged(Element* beforeChange, ChildChange::Type changeType)
+// Specs: https://html.spec.whatwg.org/multipage/dom.html#auto-directionality
+static std::optional<TextDirection> computeTextDirectionOfFormControlElement(const HTMLTextFormControlElement& textFormControlElement)
```

Comment:
I think should just merge this back into `computeAutoTextDirection() const`. It's only used once in that function and I don't see much value in splitting it.

Reply:
Merged.